src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -72,18 +72,14 @@ private static IArgumentConstraint CreateEqualityConstraint(object expressionVal
             return new EqualityArgumentConstraint(expressionValue);
         }
 
-        private static object InvokeExpression(Expression expression)
-        {
-            return Expression.Lambda(expression).Compile().DynamicInvoke();
-        }
-
         private IArgumentConstraint GetArgumentConstraintFromExpression(Expression expression)
         {
             object expressionValue = null;
 
             var trappedConstraints = this.argumentConstraintTrapper.TrapConstraints(() =>
             {
-                expressionValue = InvokeExpression(expression);
+                // in this ParsedArgumentExpression scenario, we get benefits from the same optimizations as ParsedCallExpression
+                expressionValue = ParsedCallExpression.FastEvaluate(expression);

```

Comment:
@thomaslevesque

I think in general it's such a correct evaluator that it could be used anywhere correctly, but the performance effect of optimization is potentially sensitive to the scenarios. Having it hidden away and hard to find doesn't seem a bad place to start!

I see that mainly FastEvaluate is fed various expressions that are expressions inside the lambda expression parameters of A.CallTo(). What your actual expressions look like here and therefore the arguments flowing to fastEvaluate will depend on your system under test, and test framework, however, it seems often that you are configuring a fake which is stored in a local variable, or as a property/field of some class. I also notice in A.CallTo() argument specifications it's common to do little 'ignored' argument expressions like 'A<string>._, A<blarg>.Ignored' which are also very simple expressions that this optimizes the evaluation of.

I found one or two other places in the code that expressions are evaluted using compiled lambdas - but it seemed I was hardly even hitting those code paths, and so I didn't seriously consider them yet. 

Such other spots in FakeItEasy's code:
-GetConstructorArgumentsFromExpression (another possibly good candidate if you like doing lots of constructor arguments to your fakes)
-ParseInvocationExpression (perhaps useful if you like to evaluate other lambdas as part of your A.CallTo() statements?)
-ParsedArgumentExpression.get_Value (property getter, called as part of GetArgumentConstraint for 'out' and 'ref' arguments (sometimes redundantly with getting the value as part of getting the value as part for GetArgumentConstraintFromExpression))
-and some of the FakeItEasy test code :-)